Mathlib/Tactic/MoveAdd.lean

```diff
@@ -0,0 +1,448 @@
+* the lists of pairs of a matched subexpression with `true` for an arrow `←` and `false` otherwise;
+* an array of error messages;
+* an array of debugging messages.
+
+E.g. convert `[a, ← b, _ * (1 : ℤ)]` to `pairs = ([(a, false), (z * 1, false)], , [(b, true)]`, if
+`b` does not match a subexpression of `tgt` and the first summand involving a multiplication by `1`
+is `z * 1`.
+-/
+def parsing (rws : TSyntax `Lean.Parser.Tactic.rwRuleSeq) (op : Name) (tgt : Expr) :
+    TermElabM (List (Expr × Bool) × (List MessageData × List Syntax) × Array MessageData) :=
+do match rws with
+  | `(rwRuleSeq| [$rs,*]) => do
+    let pairs := ← rs.getElems.mapM fun rstx => do
+      let r : Syntax := rstx
+      return (← Term.elabTerm r[1]! none, ! r[0]!.isNone, rstx)
+    let ops := ← getOps op tgt
+    let atoms := ((ops.map Prod.fst).foldl (· ++ ·) #[]).toList.filter ((!isBVar ·))
+    -- `instr` are the unified user-provided terms, `neverMatched` are non-unified ones
+    let (instr, neverMatched) := ← pairUp pairs.toList atoms
+    let dbgMsg := #[m!"Matching of input variables:\n* pre-match:  {
+      pairs.map (Prod.snd ∘ Prod.snd)}\n* post-match: {instr}",
+      m!"\nMaximum number of iterations: {ops.size}"]
+    -- if there are `neverMatched` terms, return the parsed terms and the syntax
+    let errMsg := neverMatched.map fun (t, a, stx) => (if a then m!"← {t}" else m!"{t}", stx)
+    return (instr, errMsg.unzip, dbgMsg)
+  | _ => failure
+
+/--  The tactic `move_add` rearranges summands of expressions.
+Calling `move_add [a, ← b, ...]` matches `a, b,...` with summands in the main goal.
+It then moves `a` to the far right and `b` to the far left of each addition in which they appear.
+The side to which the summands are moved is determined by the presence or absence of the arrow `←`.
+
+The inputs `a, b,...` can be any terms, also with underscores.
+The tactic uses the first "new" summand that unifies with each one of the given inputs.
+
+There is a multiplicative variant, called `move_mul`.
+
+There is also a general tactic for a "binary associative commutative operation": `move_oper`.
+In this case the syntax requires providing first a term whose head symbol is the operation.
+E.g. `move_oper (0 + 0) [...]` is the same as `move_add`, while `move_oper (max 0 0) [...]`
+rearranges `max`s.
+-/
+elab (name := moveOperTac) "move_oper" "(" oper:term ")" rws:rwRuleSeq  dbg:"-debug"? : tactic => do
+  -- parse the operation
+  let op := (← Term.elabTerm oper none).getAppFn.constName
+  -- parse the list of terms
+  let (instr, (unmatched, stxs), dbgMsg) ← parsing rws op (← instantiateMVars (← getMainTarget))
+  -- prepare the various error messages
+  let finErr := if unmatched.length = 0 then .nil else
+    m!"\nErrors:\nThe terms in '{unmatched}' were not matched to any atom"
+  let _ := ← stxs.mapM (logErrorAt · "") -- underline all non-matching terms
+  let _ := ← match dbg, unmatched with
```

Comment:
Can't you just write:
```suggestion
  stxs.mapM (logErrorAt · "") -- underline all non-matching terms
  match dbg, unmatched with
```

Reply:
In any case, with the `trace` this is all much cleaner.  I still could not get rid of the `let` for the `logError`, but the rest looks much tidier!